core/frontend/src/BriefcaseConnection.ts

```diff
@@ -15,14 +15,144 @@ import { IModelConnection } from "./IModelConnection";
+      for (const change of changes) {
+        this._deletedModels.delete(change.id);
+        this._modelIdToGuid.set(change.id, change.guid);
+      }
+    }));
+
+    // Buffer deletions of models.
+    this._removals.push(briefcase.txns.onModelsChanged.addListener((changes) => {
+      if (changes.deleted) {
+        for (const id of CompressedId64Set.iterable(changes.deleted)) {
+          this._modelIdToGuid.delete(id);
+          this._deletedModels.add(id);
+        }
+      }
+    }));
+
+    // Outside of an editing scope, we want to update viewport contents after commit, undo/redo, or merging changes.
+    const maybeProcess = async () => {
+      if (this.editingScope)
+        return;
+
+      const modelIds = Array.from(this._modelIdToGuid.keys());
+      if (modelIds.length > 0)
+        await IModelApp.tileAdmin.purgeTileTrees(this._briefcase, modelIds);
+
+      this.processBuffered();
+    };
+
+    this._removals.push(briefcase.txns.onCommitted.addListener(maybeProcess));
+    this._removals.push(briefcase.txns.onAfterUndoRedo.addListener(maybeProcess));
+    this._removals.push(briefcase.txns.onChangesPulled.addListener(maybeProcess));
+  }
+
+  public async close(): Promise<void> {
+    for (const removal of this._removals)
+      removal();
+
+    this._removals.length = 0;
+
+    if (this._editingScope) {
+      await this._editingScope.exit();
+      this._editingScope = undefined;
+    }
+  }
+
+  public get editingScope(): GraphicalEditingScope | undefined {
+    return this._editingScope;
+  }
+
+  public async enterEditingScope(): Promise<GraphicalEditingScope> {
+    if (this._editingScope)
+      throw new Error("Cannot create an editing scope for an iModel that already has one");
+
+    this._editingScope = await GraphicalEditingScope.enter(this._briefcase);
+
+    const removeGeomListener = this._editingScope.onGeometryChanges.addListener((changes) => {
+      const modelIds = [];
+      for (const change of changes)
+        modelIds.push(change.id);
+
+      this.invalidateScenes(modelIds);
+    });
+
+    this._editingScope.onExited.addOnce((scope) => {
+      assert(scope === this._editingScope);
+      this._editingScope = undefined;
+      removeGeomListener();
+      this.processBuffered();
+    });
+
+    return this._editingScope;
+  }
+
+  private processBuffered(): void {
+    const models = this._briefcase.models;
+    for (const [id, guid] of this._modelIdToGuid) {
+      const model = models.getLoaded(id)?.asGeometricModel;
+      if (model)
+        model.geometryGuid = guid;
+    }
+
+    const modelIds = new Set<string>(this._modelIdToGuid.keys());
+    for (const deleted of this._deletedModels) {
+      modelIds.add(deleted);
+      models.unload(deleted);
+    }
+
+    this.invalidateScenes(modelIds);
+    disposeTileTreesForGeometricModels(modelIds, this._briefcase);
+
+    this._briefcase.onBufferedModelChanges.raiseEvent(modelIds);

```

Comment:
should you send this event out after you've cleared the lists?

Reply:
Unclear to me what difference that would make - can you elaborate?

Reply:
what happens if something happens in the event that adds to those lists?

Reply:
The event is used strictly for tests, which do nothing but verify the list of model Ids is as expected.
```
  /** Strictly for tests - dispatched from ModelChangeMonitor.processBuffered.
   * @internal
   */
  public readonly onBufferedModelChanges = new BeEvent<(changedModelIds: Set<string>) => void>();
```